crates/configuration/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+        .expect("Failed to find sncast config in snfoundry.toml file");
+
+    match config.get(profile_name) {
+        Some(profile_value) => Ok(profile_value.clone()),
+        None if profile_name == "default" => Ok(Value::Table(Default::default())),
+        None => Err(anyhow!("Profile [{}] not found in config", profile_name)),
+    }
+}
+
+pub fn load_global_config<T: GlobalConfig + Default>(
+    path: &Option<Utf8PathBuf>,
+    profile: &Option<String>,
+) -> Result<T> {
+    let config_path = path
+        .as_ref()
+        .and_then(|p| search_config_upwards_relative_to(p).ok())
+        .or_else(|| find_config_file().ok());
+
+    match config_path {
+        Some(path) => {
+            let raw_config = fs::read_to_string(path)
+                .expect("Failed to read snfoundry.toml config file")
+                .parse::<Value>()
+                .expect("Failed to parse snfoundry.toml config file");
+
+            let profile = get_profile(&raw_config, T::tool_name(), profile)?;
+            T::from_raw(&profile)
+        }
+        None => Ok(T::default()),
+    }
+}
+
+pub fn search_config_upwards_relative_to(current_dir: &Utf8PathBuf) -> Result<Utf8PathBuf> {
+    current_dir
+        .ancestors()
+        .find(|path| fs::metadata(path.join(CONFIG_FILENAME)).is_ok())
+        .map(|path| path.join(CONFIG_FILENAME))
+        .ok_or_else(|| {
+            anyhow!(
+                "Failed to find snfoundry.toml - not found in current nor any parent directories"
+            )
+        })
+}
+
+pub fn find_config_file() -> Result<Utf8PathBuf> {
+    search_config_upwards_relative_to(&Utf8PathBuf::try_from(
+        std::env::current_dir().expect("Failed to get current directory"),
+    )?)
+}
+
+pub trait PropertyFromCastConfig: Sized {
+    fn from_toml_value(value: &Value) -> Option<Self>;
+}
+
+impl PropertyFromCastConfig for String {
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        value.as_str().map(std::borrow::ToOwned::to_owned)
+    }
+}
+
+impl PropertyFromCastConfig for Utf8PathBuf {
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        value.as_str().map(Utf8PathBuf::from)
+    }
+}
+
+impl PropertyFromCastConfig for u8 {
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        value.as_integer().and_then(|i| i.try_into().ok())
+    }
+}
+
+impl PropertyFromCastConfig for u16 {
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        value.as_integer().and_then(|i| i.try_into().ok())
+    }
+}
+
+impl<T> PropertyFromCastConfig for Option<T>
+where
+    T: PropertyFromCastConfig,
+{
+    fn from_toml_value(value: &Value) -> Option<Self> {
+        T::from_toml_value(value).map(Some)
+    }
+}
+
+pub fn get_property<T>(entries: &Value, field: &str) -> Option<T>
+where
+    T: PropertyFromCastConfig + Default,
+{
+    entries.get(field).and_then(T::from_toml_value)
+}
+
+#[must_use]
+pub fn copy_config_to_tempdir(src_path: &str, additional_path: Option<&str>) -> TempDir {
```

Comment:
It's only used for tests

Reply:
Yes but in different modules :(
I wanted to move it to test module but that actually is not trivial bcs it is used in multiple places. This is a reason behind duplicate I have left.